### PR TITLE
Added option to transform the paths within a sourcemap.

### PR DESCRIPTION
**Pull request by sebinsua**
This PR contains:
- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevent issue numbers:

- #2168 

### Description

> **@pgn-vole** wrote:
> It doesn't seem possible at the moment to customize the path the sources within the sourcemap.

I added a `OutputOption#sourcemapPathTransform` function which can be used to transform the source paths outputted into the sourcemaps. This helps in situations in which multiple rollup generated libraries sourcemaps are being consumed in a browser and you do not want all package's files to be grouped within the same `src/` directory. 

I do not think it's necessary to supply a base directory as was mentioned in the original GitHub issue. A transformation function can also handle making the path relative to a base directory. If you're able to pass a transform function in through a `rollup.config.js` you will also be able to pass a base path into your transformation function using a closure.

What do others think?

/cc @pgn-vole @teohhanhui

### CODE REVIEW

**Comment by sebinsua** [comment]
Can anybody help me work out why the AppVeyor build isn't working?

**Comment by teohhanhui** [comment]
> I do not think it's necessary to supply a base directory as was mentioned in the original GitHub issue. A transformation function can also handle making the path relative to a base directory. If you're able to pass a transform function in through a rollup.config.js you will also be able to pass a base path into your transformation function using a closure.

True, if they're all absolute paths, then it's fine.

**Comment by sebinsua** [comment]
> True, if they're all absolute paths, then it's fine.

Can you explain what you mean (or what you're trying to achieve), because in the unit test within this PR I have a relative path which I can then apply a transformation to.

*e.g.*

```js
const path = require('path');
const assert = require('assert');
const pkg = require('./package.json');

module.exports = {
	description: 'transform sourcemap paths (#2168)',
	options: {
		output: {
			name: 'myModule',
			file: path.resolve(__dirname, '_actual', 'bundle.js'),
			sourcemapPathTransform: sourcePath => sourcePath.replace(
                                // sourcePath is a relative path and I am doing a string-replace on it as an example. 
				`..${path.sep}`,
                                // In this example, I'm replacing it with the package name followed by a slash.
				`${pkg.name}${path.sep}`
			)
		}
	},
	test(code, map) {
		assert.deepEqual(map.sources, ['~/pkg-name/main.js']);
	}
};
```

**Comment by teohhanhui** [comment]
Sorry, I'm not familiar with how it works in rollup. But do see my gulp example in the previous issue to understand what I mean...

Anyway, as long as all the paths are already normalized to be relative to the same base directory, there's no problem.

**Comment by pgn-vole** [comment]
Looks good @sebinsua!
Thanks for that!

File: src/utils/mergeOptions.ts [hunk 1]
```diff
@@ -128,11 +128,14 @@ export default function mergeOptions({
...
 
+	const validCliOutputOptions = validOutputOptions.filter(
+		option => option !== 'sourcemapPathTransform'
+	);
 	addUnknownOptionErrors(
 		unknownOptionErrors,
 		Object.keys(command),
 		validInputOptions.concat(
-			validOutputOptions,
+			validCliOutputOptions,
```
Comment (lukastaegert) [inline src/utils/mergeOptions.ts]: Good thinking!

**Comment by lukastaegert** [comment]
One note: Once this is released, it should be documented on the website. As this tends to be forgotten and we are somewhat short-staffed, if someone would open a PR on https://github.com/rollup/rollupjs.org adding this to the big list of options (maybe with a simple example where all files paths are changed to a different base path?) this would be a great help and I will merge it once we release.

Thanks!

**Pull request merged**